overlord/snapstate/policy/snapd.go

```diff
@@ -0,0 +1,59 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2019 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package policy
+
+import (
+	"github.com/snapcore/snapd/overlord/snapstate"
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/release"
+	"github.com/snapcore/snapd/snap"
+)
+
+type snapdPolicy struct {
+	modelBase string
+}
+
+func (p *snapdPolicy) CanRemove(st *state.State, snapst *snapstate.SnapState, rev snap.Revision) error {
+	name := snapst.InstanceName()
+	if name == "" {
+		// not installed, or something. What are you even trying to do.
+		return errNoName
+	}
+
+	if !rev.Unset() {
+		return nil
+	}
+
+	// snapd cannot be removed on core
+	if !release.OnClassic {
+		return errNotRemovable
+	}
+
+	// only allow snapd removal if its the last snap on a (classic) system
+	numSnaps, err := snapstate.NumSnaps(st)
```

Comment:
_can_ you get there, on classic? You'd have to be able to remove the base first... and the model base is not removable...?

Reply:
Yes, classic is fine with this. It is possible to remove bases if they are unused. We disallow to rmeove core but that could be considered a bug I think, we should probably allow it on classic just like we do for bases.

Reply:
I just double-checked and indeed it all works.
yeah, agreed about it being a bug for core, we can probably fix that soon though
we need to find a way to restart snapd after removing the snapd snap, probably, but that's not for this PR